### PR TITLE
Add tracing with jaeger

### PR DESCRIPTION
**Pull request by fmassot**
This PR works with a jaeger instance that you can start with the following command:
`docker run -d -p6831:6831/udp -p6832:6832/udp -p16686:16686 jaegertracing/all-in-one:latest`

## Search tracing

Capture d’écran 2021-08-24 à 09 56 14

## Search stream tracing

Capture d’écran 2021-08-23 à 21 18 41

### Comments
TODO

### CODE REVIEW

**Comment by evanxg852000** [comment]
Oh nice @fmassot I was exploring and trying the integration of this on https://github.com/quickwit-inc/quickwit/issues/368
This is what I wanted to propose on this time stuff.

File: quickwit-search/src/client.rs [hunk 1]
```diff
@@ -28,11 +29,26 @@ use tokio_stream::wrappers::UnboundedReceiverStream;
...
 
 use crate::error::parse_grpc_error;
 use crate::SearchError;
 use crate::SearchService;
 
+struct MetadataMap<'a>(&'a mut tonic::metadata::MetadataMap);
+
+impl<'a> Injector for MetadataMap<'a> {
+    /// Set a key and value in the MetadataMap.  Does nothing if the key or value are not valid inputs
+    fn set(&mut self, key: &str, value: String) {
```
Comment (guilload) [inline quickwit-search/src/client.rs:43]: `value` could be a `&str`

File: quickwit-search/src/client.rs [hunk 2]
```diff
@@ -28,11 +29,26 @@ use tokio_stream::wrappers::UnboundedReceiverStream;
...
+        if let Ok(key) = tonic::metadata::MetadataKey::from_bytes(key.as_bytes()) {
```
Comment (guilload) [inline quickwit-search/src/client.rs:44]: Variable shadowing is not going to make everybody happy here :) Let' go with something along those lines:
```rust
    fn set(&mut self, key: &str, value: &str) {
        if let Ok(metadata_key) = tonic::metadata::MetadataKey::from_bytes(key.as_bytes()) {
            if let Ok(metadata_value) = tonic::metadata::MetadataValue::from_str(value) {
                self.0.insert(metadata_key, metadata_value);
            }
        }
```

File: quickwit-search/src/root.rs [hunk 3]
```diff
@@ -113,6 +120,35 @@ async fn execute_search(
     Ok(result_per_node_addr)
 }
 
+async fn execute_fetch_docs(
+    search_request: &SearchRequest,
+    partial_hits_map: &HashMap<String, Vec<PartialHit>>,
+    doc_fetch_jobs: &[(SearchServiceClient, Vec<Job>)],
+) -> anyhow::Result<Vec<Result<FetchDocsResult, SearchError>>> {
+    // Perform the fetch docs phase.
+    let mut fetch_docs_handles: Vec<JoinHandle<Result<FetchDocsResult, SearchError>>> = Vec::new();
```
Comment (guilload) [inline quickwit-search/src/root.rs:129]: I know this is not related to this PR, but we don't need to allocate a vector here: `futures::future::try_join_all` can resolve an iterator of futures.

File: quickwit-serve/src/grpc_adapter/search_adapter.rs [hunk 4]
```diff
@@ -22,13 +22,36 @@
...
 
 use quickwit_proto::{
     search_service_server as grpc, LeafSearchStreamRequest, LeafSearchStreamResult,
 };
 use quickwit_search::{SearchError, SearchService, SearchServiceImpl};
 
+struct MetadataMap<'a>(&'a tonic::metadata::MetadataMap);
+
+impl<'a> Extractor for MetadataMap<'a> {
+    /// Get a value for a key from the MetadataMap.  If the value can't be converted to &str, returns None
```
Comment (guilload) [inline quickwit-serve/src/grpc_adapter/search_adapter.rs:38]:
```suggestion
    /// Gets a value for a key from the MetadataMap. If the value can't be converted to a &str, returns None
```

**Review by guilload — COMMENTED** [review]
Left a few comments. Can't really speak for the Jaeger part of the PR since I'm unfamiliar with the framework and its API.

**Comment by fulmicoton** [comment]
Do we get the object store fetch operations too?

**Comment by fmassot** [comment]
> Do we get the object store fetch operations too?

Yes, I recorded the fetch operation at the storage directory level: we can record the file path and size to get the full picture.

Capture d’écran 2021-08-25 à 14 18 33

File: quickwit-search/src/client.rs (hunk 1 above)
Comment (fmassot) [reply quickwit-search/src/client.rs:43]: actually the trait `Injector` expects a `String`.

File: quickwit-search/src/client.rs (hunk 2 above)
Comment (fmassot) [reply quickwit-search/src/client.rs:44]: indeed I don't like it either.

**Comment by fmassot** [comment]
@fulmicoton I udpated the branch. The tracing is working for search stream only, I want to merge the retry PR before fixing tracing for search.

**Comment by fmassot** [comment]
@fulmicoton I've just rebased the branch, everything seems to work. You can set the log a debug level if you want to have information about the reading operations.

**Comment by fulmicoton** [comment]
It worked right out of the box!

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**